README
This will be written one day.


ChangeLog
2008-08-11  Eitan Isaacson  <eitan@ascender.com>

	* pyia/event.py: All sorts of crap!

2008-07-18  Eitan Isaacson  <eitan@ascender.com>

	* pyia/accessible.py: 
	Fixed state names method. 
	Make  __str__ explicityle unicode (for readability).
	Fixed special role names.

	* pyia/utils.py: 
	Added getForegroundWindow().

2008-06-23  Eitan Isaacson  <eitan@ascender.com>

	* First entry!



pyia/event.py
'''
Useful event wrappers that make life easier.

Inspired by pyatspi:
http://live.gnome.org/GAP/PythonATSPI

@author: Eitan Isaacson
@copyright: Copyright (c) 2008, Eitan Isaacson
@license: LGPL

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public
License as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.

You should have received a copy of the GNU Library General Public
License along with this library; if not, write to the
Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.
'''

from constants import winEventIDsToEventNames
from utils import accessibleObjectFromEvent

class Event(object):
    def __init__(self, 
                 event_type, hwnd, object_id, child_id, thread_id, timestamp):
        self.type = event_type
        self.hwnd = hwnd
        self.object_id = object_id
        self.child_id = child_id
        self.thread_id = thread_id
        self.timestamp = timestamp

    def __str__(self):
        return '''\
%s
\tsource: %s
\twindow: %s
\tthread: %s
\ttstamp: %s''' % \
            (winEventIDsToEventNames.get(self.type, self.type), 
             self.source, self.hwnd, self.thread_id, self.timestamp)

    def _get_source(self):
        try:
            rv = self._source
        except AttributeError:
            rv = accessibleObjectFromEvent(self)
        return rv
    
    source = property(_get_source)


pyia/utils.py
'''
A collection of useful functions to use in MSAA clients.

Inspired by pyatspi:
http://live.gnome.org/GAP/PythonATSPI

@author: Eitan Isaacson
@copyright: Copyright (c) 2008, Eitan Isaacson
@license: LGPL

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public
License as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.

You should have received a copy of the GNU Library General Public
License along with this library; if not, write to the
Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.
'''

import constants
from ctypes import windll, oledll, POINTER, byref, c_int
from comtypes.automation import VARIANT
from comtypes.gen.Accessibility import IAccessible

def getDesktop():
  desktop_hwnd = windll.user32.GetDesktopWindow()
  desktop_window = accessibleObjectFromWindow(desktop_hwnd)
  for child in desktop_window:
    if child.accRole() == constants.ROLE_SYSTEM_CLIENT:
      return child
  return None

def getForegroundWindow():
  return accessibleObjectFromWindow(
    windll.user32.GetForegroundWindow())

def accessibleObjectFromWindow(hwnd):
  ptr = POINTER(IAccessible)()
  res = oledll.oleacc.AccessibleObjectFromWindow(
    hwnd,0,
    byref(IAccessible._iid_),byref(ptr))
  return ptr

def accessibleObjectFromEvent(event):
  if not windll.user32.IsWindow(event.hwnd):
    return None
  ptr = POINTER(IAccessible)()
  varChild = VARIANT()
  res = windll.oleacc.AccessibleObjectFromEvent(
    event.hwnd, event.object_id, event.child_id,
    byref(ptr), byref(varChild))
  if res == 0:
    child=varChild.value
    return ptr.QueryInterface(IAccessible)
  else:
    return None


def findDescendant(acc, pred, breadth_first=False):
  '''
  Searches for a descendant node satisfying the given predicate starting at 
  this node. The search is performed in depth-first order by default or
  in breadth first order if breadth_first is True. For example,
  
  my_win = findDescendant(lambda x: x.name == 'My Window')
  
  will search all descendants of x until one is located with the name 'My
  Window' or all nodes are exausted. Calls L{_findDescendantDepth} or
  L{_findDescendantBreadth} to start the recursive search.
  
  @param acc: Root accessible of the search
  @type acc: Accessibility.Accessible
  @param pred: Search predicate returning True if accessible matches the 
  search criteria or False otherwise
  @type pred: callable
  @param breadth_first: Search breadth first (True) or depth first (False)?
  @type breadth_first: boolean
  @return: Accessible matching the criteria or None if not found
  @rtype: Accessibility.Accessible or None
  '''
  if breadth_first:
    return _findDescendantBreadth(acc, pred)
  
  for child in acc:
    try:
      ret = _findDescendantDepth(acc, pred)
    except Exception:
      ret = None
    if ret is not None: return ret

def _findDescendantBreadth(acc, pred):
  '''  
  Internal function for locating one descendant. Called by L{findDescendant} to
  start the search.
 
  @param acc: Root accessible of the search
  @type acc: Accessibility.Accessible
  @param pred: Search predicate returning True if accessible matches the 
  search criteria or False otherwise
  @type pred: callable
  @return: Matching node or None to keep searching
  @rtype: Accessibility.Accessible or None
  '''
  for child in acc:
    try:
      if pred(child): return child
    except Exception:
      pass
  for child in acc:
    try:
      ret = _findDescendantBreadth(child, pred)
    except Exception:
      ret = None
    if ret is not None: return ret

def _findDescendantDepth(acc, pred):
  '''
  Internal function for locating one descendant. Called by L{findDescendant} to
  start the search.
  
  @param acc: Root accessible of the search
  @type acc: Accessibility.Accessible
  @param pred: Search predicate returning True if accessible matches the 
  search criteria or False otherwise
  @type pred: callable
  @return: Matching node or None to keep searching
  @rtype: Accessibility.Accessible or None
  '''
  try:
    if pred(acc): return acc
  except Exception:
    pass
  for child in acc:
    try:
      ret = _findDescendantDepth(child, pred)
    except Exception:
      ret = None
    if ret is not None: return ret
   
def findAllDescendants(acc, pred):
  '''
  Searches for all descendant nodes satisfying the given predicate starting at 
  this node. Does an in-order traversal. For example,
  
  pred = lambda x: x.getRole() == pyatspi.ROLE_PUSH_BUTTON
  buttons = pyatspi.findAllDescendants(node, pred)
  
  will locate all push button descendants of node.
  
  @param acc: Root accessible of the search
  @type acc: Accessibility.Accessible
  @param pred: Search predicate returning True if accessible matches the 
      search criteria or False otherwise
  @type pred: callable
  @return: All nodes matching the search criteria
  @rtype: list
  '''
  matches = []
  _findAllDescendants(acc, pred, matches)
  return matches

def _findAllDescendants(acc, pred, matches):
  '''
  Internal method for collecting all descendants. Reuses the same matches
  list so a new one does not need to be built on each recursive step.
  '''
  for child in acc:
    try:
      if pred(child): matches.append(child)
    except Exception:
      pass
    _findAllDescendants(child, pred, matches)

def printSubtree(acc, indent=0):
  print '%s%s' % (indent*' ', unicode(acc).encode('cp1252', 'ignore'))
  for child in acc:
    try:
      printSubtree(child, indent+1)
    except:
      pass

def windowFromAccessibleObject(acc):
  hwnd = c_int()
  try:
    res = windll.oleacc.WindowFromAccessibleObject(acc, byref(hwnd))
  except:
    res = 0
  if res == 0:
    return hwnd.value
  else:
    return 0

def getWindowThreadProcessID(hwnd):
  processID = c_int()
  threadID = windll.user32.GetWindowThreadProcessId(hwnd,byref(processID))
  return (processID.value, threadID)

def getAccessibleThreadProcessID(acc):
  hwnd = windowFromAccessibleObject(acc)
  return getWindowThreadProcessID(hwnd)


pyia/__init__.py
'''
A Python MSAA client libary.

Inspired by pyatspi:
http://live.gnome.org/GAP/PythonATSPI

@author: Eitan Isaacson
@copyright: Copyright (c) 2008, Eitan Isaacson
@license: LGPL

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public
License as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.

You should have received a copy of the GNU Library General Public
License along with this library; if not, write to the
Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.
'''

__version__ = "0.0.1"
from comtypes.client import GetModule
GetModule('oleacc.dll')
from comtypes.gen.Accessibility import IAccessible
del GetModule
import accessible
from utils import *
from constants import *
import registry

# Create singleton registry.
Registry = registry.Registry()
registry.Registry = Registry
del registry


pyia/registry.py
'''
A registry, sort of like AT-SPI has. For now it's just an entry point for 
registering liteners for MSAA events.

Inspired by pyatspi:
http://live.gnome.org/GAP/PythonATSPI

@author: Eitan Isaacson
@copyright: Copyright (c) 2008, Eitan Isaacson
@license: LGPL

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public
License as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.

You should have received a copy of the GNU Library General Public
License along with this library; if not, write to the
Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.
'''

import constants
from ctypes import CFUNCTYPE, c_int, c_voidp, windll
from comtypes.client import PumpEvents
from event import Event
from utils import accessibleObjectFromEvent

class Registry(object):
    def __init__(self):
        self.clients = {}
        self.hook_ids = []
        self._c_handleEvent = CFUNCTYPE(
                c_voidp,c_int,c_int,c_int,c_int,c_int,c_int,c_int)(
                    self._handleEvent)

    def __call__(self):
        return self

    def _handleEvent(self, handle, eventID, window, objectID, childID, 
                     threadID, timestamp):
        e = Event(eventID, window, objectID, childID, threadID, timestamp)
        for client, event_type in self.clients:
            if event_type == eventID:
                client(e)
            
    def registerEventListener(self, client, *event_types):
        for event_type in event_types:
            if self.clients.has_key((client, event_type)):
                continue
            hook_id = \
                windll.user32.SetWinEventHook(
                    event_type, event_type, 0, self._c_handleEvent, 0, 0, 0)
            if hook_id:
                self.clients[(client, event_type)] = hook_id
            else:
                print "Could not register callback for %s" % \
                    constants.winEventIDsToEventNames.get(event_type, event_type)

    def deregisterEventListener(self, client, *event_types):
        for event_type in event_types:
            try:
                windll.user32.UnhookWinEvent(
                    self.clients[(client, event_type)])
                del self.clients[(client, event_type)]
            except KeyError:
                pass

    def clearListeners(self):
        while True:
            try:
                client_tuple, hook_id = self.clients.popitem()
            except KeyError:
                break
            else:
                windll.user32.UnhookWinEvent(hook_id)


    def iter_loop(self, timeout=1):
        PumpEvents(timeout)
        
    def start(self):
        while True:
            try:
                self.iter_loop(5)
            except KeyboardInterrupt:
                self.clearListeners()
                break


pyia/accessible.py
'''
Creates functions at import time that are mixed into the 
IAccessible base class to make it more Pythonic.

Inspired by pyatspi:
http://live.gnome.org/GAP/PythonATSPI

@author: Eitan Isaacson
@copyright: Copyright (c) 2008, Eitan Isaacson
@license: LGPL

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public
License as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.

You should have received a copy of the GNU Library General Public
License along with this library; if not, write to the
Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.
'''

import new
import types
from comtypes.automation import VARIANT, VT_I4, VT_DISPATCH
from ctypes import c_long, oledll, byref, create_unicode_buffer
from comtypes.gen.Accessibility import IAccessible
from comtypes import named_property, COMError, hresult
from constants import CHILDID_SELF

def _makeExceptionHandler(func):
    '''
    Builds a function calling the one it wraps in try/except statements catching
    COMError exceptions.
  
    @return: Function calling the method being wrapped
    @rtype: function
    '''
    def _inner(self, *args, **kwargs):
        try:
            return func(self, *args, **kwargs)
        except COMError, e:
            # TODO: Translate COMErrors to more pythonic equivalents.
            raise
    return _inner

def _mixExceptions(cls):
    '''
    Wraps all methods and properties in a class with handlers for CORBA 
    exceptions.
    
    @param cls: Class to mix interface methods into
    @type cls: class
    '''
    # get a method type as a reference from a known method
    # loop over all names in the new class
    for name in cls.__dict__.keys():
        obj = cls.__dict__[name]
        # check if we're on a protected or private method
        if name.startswith('_'):
            continue
        # check if we're on a method
        elif isinstance(obj, new.instancemethod):
            # wrap the function in an exception handler
            method = _makeExceptionHandler(obj)
            # add the wrapped function to the class
            setattr(cls, name, method)
        elif isinstance(obj, named_property):
            # wrap the function in an exception handler
            if obj.getter is not None:
                obj.getter = _makeExceptionHandler(obj.getter)
            if obj.setter is not None:
                obj.setter = _makeExceptionHandler(obj.setter)
        # check if we're on a property
        elif isinstance(obj, property):
            # wrap the getters and setters
            if obj.fget and obj.fget.__name__ != '_inner':
                getter = _makeExceptionHandler(obj.fget)
            else:
                getter = None
            if obj.fset and obj.fset.__name__ != '_inner':
                setter = _makeExceptionHandler(obj.fset)
            else:
                setter = None
            setattr(cls, name, property(getter, setter))

def _mixClass(cls, new_cls, ignore=[]):
    '''
    Adds the methods in new_cls to cls. After mixing, all instances of cls will
    have the new methods. If there is a method name clash, the method already 
    in cls will be prefixed with '_mix_' before the new method of the same 
    name is mixed in.
    
    @note: _ is not the prefix because if you wind up with __ in front of a 
    variable, it becomes private and mangled when an instance is created. 
    Difficult to invoke from the mixin class.
    
    @param cls: Existing class to mix features into
    @type cls: class
    @param new_cls: Class containing features to add
    @type new_cls: class
    @param ignore: Ignore these methods from the mixin
    @type ignore: iterable
    '''
    # loop over all names in the new class
    for name, func in new_cls.__dict__.items():
        if name in ignore:
            continue
        if isinstance(func, types.FunctionType):
            # build a new function that is a clone of the one from new_cls
            method = new.function(func.func_code, func.func_globals, name, 
                                  func.func_defaults, func.func_closure)
            try:
                # check if a method of the same name already exists in the 
                # target
                old_method = getattr(cls, name)
            except AttributeError:
                pass
            else:
                # rename the old method so we can still call it if need be
                setattr(cls, '_mix_'+name, old_method)
            # add the clone to cls
            setattr(cls, name, method)
        elif isinstance(func, staticmethod):
            try:
                # check if a method of the same name already exists 
                # in the target
                old_method = getattr(cls, name)
            except AttributeError:
                pass
            else:
                # rename the old method so we can still call it if need be
                setattr(cls, '_mix_'+name, old_method)
            setattr(cls, name, func)
        elif isinstance(func, property):
            try:
                # check if a method of the same name already exists 
                # in the target
                old_prop = getattr(cls, name)
            except AttributeError:
                pass
            else:
                # IMPORTANT: We save the old property before overwriting it, 
                # even though we never end up calling the old prop from our 
                # mixin class If we don't save the old one, we seem to 
                # introduce a Python ref count problem where the property 
                # get/set methods disappear before we can use them at a later 
                # time. This is a minor waste of memory because a property is 
                # a class object and we only overwrite a few of them.
                setattr(cls, '_mix_'+name, old_prop)
            setattr(cls, name, func)


class _IAccessibleMixin(object):
    def __getitem__(self, index):
        n = self.accChildCount
        if index >= n or index < -n:
            raise IndexError
        elif index < 0:
            index += n
        children = (VARIANT*1)()
        pcObtained = c_long()
        oledll.oleacc.AccessibleChildren(
            self, index, 1, children, byref(pcObtained))
        child = children[0]
        if child.vt == VT_I4:
            try:
                return \
                    self.accChild(child).QueryInterface(IAccessible)
            except:
                raise IndexError
        elif child.vt == VT_DISPATCH:
            return child.value.QueryInterface(IAccessible)
        else:
            for i, item in enumerate(self):
                if i == index:
                    return item
        raise IndexError


    def __iter__(self):
        accChildCount = self.accChildCount
        VariantArrayType = VARIANT * accChildCount
        rgvarChildren = VariantArrayType()
        pcObtained = c_long()
        oledll.oleacc.AccessibleChildren(self, 0, accChildCount, rgvarChildren,
                                  byref(pcObtained))
        for child in rgvarChildren:
            if child.vt == VT_I4:
                try:
                    ppdispChild = ia.accChild(child)
                    if ppdispChild:
                        yield ppdispChild.QueryInterface(IAccessible)
                except:
                    pass
            elif child.vt == VT_DISPATCH:
                yield child.value.QueryInterface(IAccessible)

    def __str__(self):
        try:
            return u'[%s | %s]' % (self.accRoleName(), 
                                   self.accName(CHILDID_SELF) or '')
        except:
            raise
            return u'[DEAD]'

    def __len__(self):
        return self.accChildCount

    def accStateSet(self):
        states = []
        state = self.accState(CHILDID_SELF)
        for shift in xrange(64):
            state_bit = 1 << shift
            if state_bit & state:
                states.append(self._getStateText(state_bit & state))
        return states

    def _getStateText(self, state_bit):
        stateLen = oledll.oleacc.GetStateTextW(state_bit,0,0)
        if stateLen:
            buf = create_unicode_buffer(stateLen + 2)
            oledll.oleacc.GetStateTextW(state_bit, buf, stateLen + 1)
            return buf.value
        else:
            return ''
        
    def accRoleName(self):
        role = self.accRole(CHILDID_SELF)
        if not isinstance(role, int):
            # Maybe one of those Mozilla string roles, just return it.
            return role
        roleLen = oledll.oleacc.GetRoleTextW(role,0,0)
        if roleLen:
            buf = create_unicode_buffer(roleLen + 2)
            oledll.oleacc.GetRoleTextW(role, buf, roleLen + 1)
            return buf.value
        else:
            return ''

_mixExceptions(IAccessible)
_mixClass(IAccessible, _IAccessibleMixin)


pyia/constants.py
'''
Useful constants.

Inspired by pyatspi:
http://live.gnome.org/GAP/PythonATSPI

@author: Eitan Isaacson
@copyright: Copyright (c) 2008, Eitan Isaacson
@license: LGPL

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public
License as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.

You should have received a copy of the GNU Library General Public
License along with this library; if not, write to the
Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.
'''
# Child ID.
CHILDID_SELF = 0
# Accessible Roles 
# TODO: Is there a way to retrieve this at runtime or build time?
#
ROLE_SYSTEM_ALERT = 8
ROLE_SYSTEM_ANIMATION = 54
ROLE_SYSTEM_APPLICATION = 14
ROLE_SYSTEM_BORDER = 19
ROLE_SYSTEM_BUTTONDROPDOWN  = 56
ROLE_SYSTEM_BUTTONDROPDOWNGRID = 58
ROLE_SYSTEM_BUTTONMENU = 57
ROLE_SYSTEM_CARET = 7
ROLE_SYSTEM_CELL = 29
ROLE_SYSTEM_CHARACTER = 32
ROLE_SYSTEM_CHART = 17
ROLE_SYSTEM_CHECKBUTTON = 44
ROLE_SYSTEM_CLIENT = 10
ROLE_SYSTEM_CLOCK = 61
ROLE_SYSTEM_COLUMN = 27
ROLE_SYSTEM_COLUMNHEADER = 25
ROLE_SYSTEM_COMBOBOX = 46
ROLE_SYSTEM_CURSOR = 6
ROLE_SYSTEM_DIAGRAM = 53
ROLE_SYSTEM_DIAL = 49
ROLE_SYSTEM_DIALOG = 18
ROLE_SYSTEM_DOCUMENT = 15
ROLE_SYSTEM_DROPLIST = 47
ROLE_SYSTEM_EQUATION = 55
ROLE_SYSTEM_GRAPHIC = 40
ROLE_SYSTEM_GRIP = 4
ROLE_SYSTEM_GROUPING = 20
ROLE_SYSTEM_HELPBALLOON = 31
ROLE_SYSTEM_HOTKEYFIELD = 50
ROLE_SYSTEM_INDICATOR = 39
ROLE_SYSTEM_LINK = 30
ROLE_SYSTEM_LIST = 33
ROLE_SYSTEM_LISTITEM = 34
ROLE_SYSTEM_MENUBAR = 2
ROLE_SYSTEM_MENUITEM = 12
ROLE_SYSTEM_MENUPOPUP = 11
ROLE_SYSTEM_OUTLINE = 35
ROLE_SYSTEM_OUTLINEITEM = 36
ROLE_SYSTEM_PAGETAB = 37
ROLE_SYSTEM_PAGETABLIST = 60
ROLE_SYSTEM_PANE = 16
ROLE_SYSTEM_PROGRESSBAR = 48
ROLE_SYSTEM_PROPERTYPAGE = 38
ROLE_SYSTEM_PUSHBUTTON = 43
ROLE_SYSTEM_RADIOBUTTON = 45
ROLE_SYSTEM_ROW = 28
ROLE_SYSTEM_ROWHEADER = 26
ROLE_SYSTEM_SCROLLBAR = 3
ROLE_SYSTEM_SEPARATOR = 21
ROLE_SYSTEM_SLIDER = 51
ROLE_SYSTEM_SOUND = 5
ROLE_SYSTEM_SPINBUTTON = 52
ROLE_SYSTEM_STATICTEXT = 41
ROLE_SYSTEM_STATUSBAR = 23
ROLE_SYSTEM_TABLE = 24
ROLE_SYSTEM_TEXT = 42
ROLE_SYSTEM_TITLEBAR = 1
ROLE_SYSTEM_TOOLBAR = 22
ROLE_SYSTEM_TOOLTIP = 13
ROLE_SYSTEM_WHITESPACE = 59
ROLE_SYSTEM_WINDOW = 9

# Navigation constants
NAVDIR_DOWN = 2
NAVDIR_FIRSTCHILD = 7
NAVDIR_LASTCHILD = 8
NAVDIR_LEFT = 3
NAVDIR_NEXT = 5
NAVDIR_PREVIOUS = 6
NAVDIR_RIGHT = 4
NAVDIR_UP = 1

STATE_SYSTEM_UNAVAILABLE = 0x1
STATE_SYSTEM_SELECTED = 0x2
STATE_SYSTEM_FOCUSED = 0x4
STATE_SYSTEM_PRESSED = 0x8
STATE_SYSTEM_CHECKED = 0x10
STATE_SYSTEM_MIXED = 0x20
STATE_SYSTEM_READONLY = 0x40
STATE_SYSTEM_HOTTRACKED = 0x80
STATE_SYSTEM_DEFAULT = 0x100
STATE_SYSTEM_EXPANDED = 0x200
STATE_SYSTEM_COLLAPSED = 0x400
STATE_SYSTEM_BUSY = 0x800
STATE_SYSTEM_FLOATING = 0x1000
STATE_SYSTEM_MARQUEED = 0x2000
STATE_SYSTEM_ANIMATED = 0x4000
STATE_SYSTEM_INVISIBLE = 0x8000
STATE_SYSTEM_OFFSCREEN = 0x10000
STATE_SYSTEM_SIZEABLE = 0x20000
STATE_SYSTEM_MOVEABLE = 0x40000
STATE_SYSTEM_SELFVOICING = 0x80000
STATE_SYSTEM_FOCUSABLE = 0x100000
STATE_SYSTEM_SELECTABLE = 0x200000
STATE_SYSTEM_LINKED = 0x400000
STATE_SYSTEM_TRAVERSED = 0x800000
STATE_SYSTEM_MULTISELECTABLE = 0x1000000
STATE_SYSTEM_EXTSELECTABLE = 0x2000000
STATE_SYSTEM_HASSUBMENU = 0x4000000
STATE_SYSTEM_ALERT_LOW = 0x4000000
STATE_SYSTEM_ALERT_MEDIUM = 0x8000000
STATE_SYSTEM_ALERT_HIGH = 0x10000000
STATE_SYSTEM_PROTECTED = 0x20000000
STATE_SYSTEM_HASPOPUP = 0x40000000
STATE_SYSTEM_VALID = 0x1fffffff

#win events
EVENT_SYSTEM_SOUND = 0x1
EVENT_SYSTEM_ALERT = 0x2
EVENT_SYSTEM_FOREGROUND = 0x3
EVENT_SYSTEM_MENUSTART = 0x4
EVENT_SYSTEM_MENUEND = 0x5
EVENT_SYSTEM_MENUPOPUPSTART = 0x6
EVENT_SYSTEM_MENUPOPUPEND = 0x7
EVENT_SYSTEM_CAPTURESTART = 0x8
EVENT_SYSTEM_CAPTUREEND = 0x9
EVENT_SYSTEM_MOVESIZESTART = 0xa
EVENT_SYSTEM_MOVESIZEEND = 0xb
EVENT_SYSTEM_CONTEXTHELPSTART = 0xc
EVENT_SYSTEM_CONTEXTHELPEND = 0xd
EVENT_SYSTEM_DRAGDROPSTART = 0xe
EVENT_SYSTEM_DRAGDROPEND = 0xf
EVENT_SYSTEM_DIALOGSTART = 0x10
EVENT_SYSTEM_DIALOGEND = 0x11
EVENT_SYSTEM_SCROLLINGSTART = 0x12
EVENT_SYSTEM_SCROLLINGEND = 0x13
EVENT_SYSTEM_SWITCHSTART = 0x14
EVENT_SYSTEM_SWITCHEND = 0x15
EVENT_SYSTEM_MINIMIZESTART = 0x16
EVENT_SYSTEM_MINIMIZEEND = 0x17
EVENT_OBJECT_CREATE = 0x8000
EVENT_OBJECT_DESTROY = 0x8001
EVENT_OBJECT_SHOW = 0x8002
EVENT_OBJECT_HIDE = 0x8003
EVENT_OBJECT_REORDER = 0x8004
EVENT_OBJECT_FOCUS = 0x8005
EVENT_OBJECT_SELECTION = 0x8006
EVENT_OBJECT_SELECTIONADD = 0x8007
EVENT_OBJECT_SELECTIONREMOVE = 0x8008
EVENT_OBJECT_SELECTIONWITHIN = 0x8009
EVENT_OBJECT_STATECHANGE = 0x800a
EVENT_OBJECT_LOCATIONCHANGE = 0x800b
EVENT_OBJECT_NAMECHANGE = 0x800c
EVENT_OBJECT_DESCRIPTIONCHANGE = 0x800d
EVENT_OBJECT_VALUECHANGE = 0x800e
EVENT_OBJECT_PARENTCHANGE = 0x800f
EVENT_OBJECT_HELPCHANGE = 0x8010
EVENT_OBJECT_DEFACTIONCHANGE = 0x8011
EVENT_OBJECT_ACCELERATORCHANGE = 0x8012
EVENT_CONSOLE_CARET = 0x4001
EVENT_CONSOLE_UPDATE_REGION = 0x4002
EVENT_CONSOLE_UPDATE_SIMPLE = 0x4003
EVENT_CONSOLE_UPDATE_SCROLL = 0x4004
EVENT_CONSOLE_LAYOUT = 0x4005
EVENT_CONSOLE_START_APPLICATION = 0x4006
EVENT_CONSOLE_END_APPLICATION = 0x4007

winEventIDsToEventNames={}

for _sym, _val in locals().items():
    if _sym.startswith('EVENT_'):
        winEventIDsToEventNames[_val] = _sym
